### PR TITLE
Raise minimum build JVM version to Java 8

### PR DESCRIPTION
**Pull request by oehme**
Doing so uncovered a few issues:

- testing a JVM version change for Scala is no longer possible as our workers don't support a custom Java home and the Scala versions we use don't support Java 9+
- the bootclasspath test was broken and didn't actually test setting the bootclasspath. As a result we missed a bug on Java 9+
- Groovy 2.4 has trouble handling bridge methods generated by Java 8, some of our test fixtures don't compile as a result. I hope this will be fixed in Groovy 2.5 (cc @lptr)

I suggest we address these separately.

### CODE REVIEW

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 1]
```diff
@@ -1219,15 +1218,6 @@ private void validate(String output, String displayName) {
                     if (line.matches(".*use(s)? or override(s)? a deprecated API\\.")) {
                         // A javac warning, ignore
                         i++;
-                    } else if (line.contains(JAVA7_DEPRECATION_WARNING.substring(0, JAVA7_DEPRECATION_WARNING.length() - 1))) {
```
Comment (blindpirate) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: The [relevant private methods](https://github.com/gradle/gradle/blob/a704411c1156968b702bb21a2b0aa4d43ae82894/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java#L1038) also need to be removed.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 2]
```diff
@@ -1219,15 +1218,6 @@ private void validate(String output, String displayName) {
...
-                        if (!java7DeprecationWarningShouldExist()) {
-                            throw new AssertionError(String.format("%s line %d contains unexpected deprecation warning: %s%n=====%n%s%n=====%n", displayName, i + 1, line, output));
-                        }
-                        // skip over stack trace
-                        i++;
-                        while (i < lines.size() && STACK_TRACE_ELEMENT.matcher(lines.get(i)).matches()) {
-                            i++;
-                        }
                     } else if (isJava7DeprecationTruncatedDaemonLog(lines, i, line)) {
```
Comment (blindpirate) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Seems like this block also needs to be removed.

**Comment by lptr** [comment]
What's the issue with the bridge methods? You can use the `lptr/core/upgrade-to-groovy-2.5` branch to try something out if you want. In any case I hope to merge it today.

**Comment by oehme** [comment]
@lptr Java 8 seems to create some additional bridge methods (which is also why our API report warns about a "new" method that isn't actually new). Groovy 2.4 reports these bridge methods as errors saying "This method has the same name but doesn't overwrite the supertype method".

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.java [hunk 3]
```diff
@@ -93,7 +93,11 @@ private boolean worksWith(JavaVersion javaVersion) {
             return javaVersion.compareTo(JavaVersion.VERSION_1_7) >= 0 && javaVersion.compareTo(JavaVersion.VERSION_1_8) <= 0;
         }
 
-        return javaVersion.compareTo(JavaVersion.VERSION_1_7) >= 0 && javaVersion.compareTo(JavaVersion.VERSION_1_10) <= 0;
+        if (isSameOrOlder("4.10")) {
```
Comment (marcphilipp) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.java]: Didn't we support Java 10 with 4.7?

Comment (oehme) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.java]: Yes, it's "isSameOrOlder"

File: subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy [hunk 4]
```diff
@@ -31,6 +32,7 @@ class UpToDatePlatformScalaCompileIntegrationTest extends AbstractIntegrationSpe
         file('app/controller/Person.scala') << "class Person(name: String)"
     }
 
+    @Ignore("TODO upgrade Scala to be Java 9+ compatible or make workers honor javaHome to enable this test again")
```
Comment (wolfs) [inline subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy]: @oehme What is the problem here? Scala works on Java 9, only incremental scala compilation does not work (i.e. the compilation is always non-incremental). So running this test with Java 8/9 should be fine, right?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.java (hunk 3 above)
Comment (marcphilipp) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.java]: Oh, nevermind then. 🙈

File: subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy (hunk 4 above)
Comment (oehme) [reply subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy]: Let me try that again, maybe it was due to some other problem that has now been fixed.

**Comment by oehme** [comment]
@marcphilipp Indeed there were some mentions of Java 7 that I updated. I also added a breaking change to the release notes.

File: subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy (hunk 4 above)
Comment (oehme) [reply subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy]: Nope, still failing: https://builds.gradle.org/viewLog.html?buildId=15657513&tab=buildResultsDiv&buildTypeId=Gradle_Check_Quick_Java10_Oracle_Linux_scala

Is there a specific Scala version that I should be avoiding?

Comment (oehme) [reply subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/UpToDatePlatformScalaCompileIntegrationTest.groovy]: It's the scala version after all: https://docs.scala-lang.org/overviews/jdk-compatibility/overview.html

This test uses 2.11.8 currently

**Comment by oehme** [comment]
@marcphilipp PTAL

**Pull request merged**